File: scripts/tasks/doctor.py
"""Health check for toolchain, core files, and managed projects."""

from __future__ import annotations

from pathlib import Path

from scripts.common import command_exists, log_step, run_or_throw
from scripts.projects_config import MANAGED_PROJECTS


def doctor() -> None:
    log_step("Toolchain status")
    for tool in ["bun", "git", "docker", "python3", "ruff", "uv"]:
        if not command_exists(tool):
            print(f"missing: {tool}")
            continue

        version = run_or_throw([tool, "--version"], quiet=True).split("\n")[0]
        print(f"{tool}: {version}")

    log_step("Core files")
    cwd = Path.cwd()
    for file in ["SOUL.md", "AGENTS.md", "README.md", "scripts/cli.py"]:
        status = "ok" if (cwd / file).exists() else "missing"
        print(f"{file}: {status}")

    log_step("Managed projects")
    if not MANAGED_PROJECTS:
        print("(none configured)")
        return

    for project in MANAGED_PROJECTS:
        has_repo = project.path.exists() and (project.path / ".git").exists()
        status = "ok" if has_repo else "missing"
        print(f"{project.name}: {status} ({project.path})")
        if not has_repo:
            continue

        branch = run_or_throw(
            ["git", "branch", "--show-current"], cwd=str(project.path), quiet=True
        )

        push_urls = run_or_throw(
            ["git", "remote", "get-url", "--all", "--push", "origin"],
            cwd=str(project.path),
            quiet=True,
        )

        urls = " | ".join(
            line.strip() for line in push_urls.split("\n") if line.strip()
        )
        print(f"branch: {branch}")
        print(f"push: {urls}")

File: packages/cli/src/commands/doctor.ts
import fs from 'node:fs'
import path from 'node:path'
import {
  currentGitBranch,
  gitRemotePushUrls,
  isGitRepo,
  logStep,
  managedProjects,
  runCommandText,
} from '@scry-home/core'

import { commandExists } from '../lib/system'

export const doctor = async () => {
  logStep('Toolchain status')
  for (const tool of ['git', 'docker', 'bun', 'node']) {
    if (!commandExists(tool)) {
      process.stdout.write(`missing: ${tool}\n`)
      continue
    }

    const version = await runCommandText([tool, '--version'], { quiet: true })
    process.stdout.write(`${tool}: ${version.split('\n')[0]}\n`)
  }

  logStep('Core files')
  const cwd = process.cwd()
  for (const file of [
    'README.md',
    'BUILD.md',
    'package.json',
    path.join('packages', 'cli', 'src', 'index.ts'),
  ]) {
    process.stdout.write(`${file}: ${fs.existsSync(path.join(cwd, file)) ? 'ok' : 'missing'}\n`)
  }

  logStep('Managed projects')
  for (const project of managedProjects) {
    const hasRepo = isGitRepo(project.path)
    process.stdout.write(`${project.name}: ${hasRepo ? 'ok' : 'missing'} (${project.path})\n`)
    if (!hasRepo) {
      continue
    }

    const branch = await currentGitBranch(project.path)
    process.stdout.write(`branch: ${branch}\n`)

    const originUrls = await gitRemotePushUrls(project.path, 'origin')
    if (originUrls.length > 0) {
      process.stdout.write(`push(origin): ${originUrls.join(' | ')}\n`)
    }

    const forkUrls = await gitRemotePushUrls(project.path, 'fork')
    if (forkUrls.length > 0) {
      process.stdout.write(`push(fork): ${forkUrls.join(' | ')}\n`)
    }
  }
}
